File: swift/AntennaProtocol/Sources/AntennaProtocol/EventID.swift
import Foundation
import CryptoKit

public enum MBEventId {
    /// Strip fields excluded from eventId hashing per MBP2P v0.1.0:
    /// - id (self-referential)
    /// - auth (variable signatures/proofs)
    /// - thread (root posts often set thread == id)
    /// - metadata (non-normative hints; avoids recursion such as replyTopic derived from id)
    public static let stripFields: Set<String> = ["id", "auth", "thread", "metadata"]

    public static func compute(from eventJSON: Any) throws -> String {
        guard var obj = eventJSON as? [String: Any] else {
            throw MBP2PError.invalidJSON("Event root must be an object")
        }
        for k in stripFields { obj.removeValue(forKey: k) }
        let canon = try JCS.canonicalize(obj)
        let digest = SHA256.hash(data: canon.data(using: .utf8)!)
        return "0x" + digest.compactMap { String(format: "%02x", $0) }.joined()
    }
}

public extension MBEvent {
    static func fromJSON(data: Data) throws -> MBEvent {
        return try MBJSON.decode(MBEvent.self, from: data)
    }

    func computeEventId() throws -> String {
        var obj = try MBJSON.toJSONObject(self)
        // Keep author optional identity keys explicit as null to match the shared
        // interop vectors and other reference implementations.
        if var eventDict = obj as? [String: Any], var author = eventDict["author"] as? [String: Any] {
            if author["agentRegistry"] == nil { author["agentRegistry"] = NSNull() }
            if author["agentId"] == nil { author["agentId"] = NSNull() }
            if author["ens"] == nil { author["ens"] = NSNull() }
            if author["anonKey"] == nil { author["anonKey"] = NSNull() }
            eventDict["author"] = author
            obj = eventDict
        }
        return try MBEventId.compute(from: obj)
    }
}

File: rust/antenna-protocol/src/event_id.rs
use crate::{error::MBP2PError, jcs, types::MBEvent};
use serde_json::Value;
use sha2::{Digest, Sha256};

/// Fields excluded from eventId hashing per MBP2P v0.1.0.
pub const STRIP_FIELDS: [&str; 4] = ["id", "auth", "thread", "metadata"];

/// Compute event.id = 0x + SHA-256( canonicalize(event without id/auth/thread/metadata) )
pub fn compute_event_id(event: &MBEvent) -> Result<String, MBP2PError> {
    let mut v = serde_json::to_value(event).map_err(|e| MBP2PError::InvalidJson(e.to_string()))?;
    strip_fields(&mut v)?;
    let canon = jcs::canonicalize(&v)?;
    let digest = Sha256::digest(canon.as_bytes());
    Ok(format!("0x{}", hex::encode(digest)))
}

fn strip_fields(v: &mut Value) -> Result<(), MBP2PError> {
    let obj = v.as_object_mut().ok_or_else(|| MBP2PError::InvalidJson("event must be object".into()))?;
    for k in STRIP_FIELDS {
        obj.remove(k);
    }
    Ok(())
}
